storage: TestReplicateQueueDownReplicate failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/2cbfb514fed209e9e4192bd07af6baa8dd073bab

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestReplicateQueueDownReplicate PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=971557&tab=buildLog

```
1      [async] transport racer
W181017 07:56:00.886909 117992 storage/raft_transport.go:583  [n3] while processing outgoing Raft queue to node 4: rpc error: code = Canceled desc = grpc: the client connection is closing:
I181017 07:56:00.892252 119984 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n4: context canceled
I181017 07:56:00.894779 119986 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n1: context canceled
W181017 07:56:00.895569 118546 storage/raft_transport.go:583  [n3] while processing outgoing Raft queue to node 2: rpc error: code = Canceled desc = grpc: the client connection is closing:
W181017 07:56:00.897266 118568 storage/raft_transport.go:583  [n2] while processing outgoing Raft queue to node 3: rpc error: code = Unavailable desc = transport is closing:
W181017 07:56:00.904696 118212 storage/raft_transport.go:583  [n2] while processing outgoing Raft queue to node 4: rpc error: code = Canceled desc = grpc: the client connection is closing:
W181017 07:56:00.906965 117981 storage/raft_transport.go:583  [n2] while processing outgoing Raft queue to node 5: rpc error: code = Canceled desc = context canceled:
W181017 07:56:00.918946 117953 storage/raft_transport.go:583  [n4] while processing outgoing Raft queue to node 3: EOF:
I181017 07:56:00.919612 120026 util/stop/stopper.go:537  quiescing; tasks left:
2      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181017 07:56:00.921530 120024 util/stop/stopper.go:537  quiescing; tasks left:
2      [async] closedts-subscription
W181017 07:56:00.936350 118405 storage/raft_transport.go:583  [n4] while processing outgoing Raft queue to node 5: EOF:
I181017 07:56:00.941820 120069 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n1: context canceled
I181017 07:56:00.944618 120025 util/stop/stopper.go:537  quiescing; tasks left:
3      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181017 07:56:00.945149 120025 util/stop/stopper.go:537  quiescing; tasks left:
3      [async] closedts-subscription
I181017 07:56:00.945586 120025 util/stop/stopper.go:537  quiescing; tasks left:
2      [async] closedts-subscription
I181017 07:56:00.948990 120083 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n4: context canceled
I181017 07:56:00.950825 119906 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n3: context canceled
I181017 07:56:00.952485 120084 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n5: context canceled
I181017 07:56:00.953458 119985 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n5: context canceled
I181017 07:56:00.954696 120068 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n5: context canceled
I181017 07:56:00.957513 120023 util/stop/stopper.go:537  quiescing; tasks left:
3      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181017 07:56:00.957776 120085 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n1: context canceled
W181017 07:56:00.959279 118159 storage/raft_transport.go:583  [n4] while processing outgoing Raft queue to node 2: rpc error: code = Canceled desc = grpc: the client connection is closing:
I181017 07:56:00.959331 120067 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n3: context canceled
I181017 07:56:00.959536 120023 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181017 07:56:00.960265 120023 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
I181017 07:56:01.073252 119920 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n2: context canceled
I181017 07:56:01.085230 115555 kv/transport_race.go:113  transport race promotion: ran 54 iterations on up to 3113 requests
I181017 07:56:01.086536 120103 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n2: context canceled
--- FAIL: TestReplicateQueueDownReplicate (32.09s)
	testing.go:730: race detected during execution of test

```

--- a/Makefile
+++ b/Makefile
@@ -838,7 +838,7 @@ stressrace: ## Run tests under stress with the race detector enabled.
 stress stressrace:
 	$(xgo) test $(GOFLAGS) -exec 'stress $(STRESSFLAGS)' -tags '$(TAGS)' -ldflags '$(LINKFLAGS)' -run "$(TESTS)" -timeout 0 $(PKG) $(filter-out -v,$(TESTFLAGS)) -v -args -test.timeout $(TESTTIMEOUT)
 
-.PHONE: roachprod-stress roachprod-stressrace
+.PHONY: roachprod-stress roachprod-stressrace
 roachprod-stress roachprod-stressrace: bin/roachprod-stress
 	# The bootstrap target creates, among other things, ./bin/stress.
 	build/builder.sh make bin/.bootstrap

--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -870,13 +870,14 @@ func (r *Replica) destroyRaftMuLocked(ctx context.Context, nextReplicaID roachpb
 
 func (r *Replica) cancelPendingCommandsLocked() {
 	r.mu.AssertHeld()
-	pr := proposalResult{
-		Err:           roachpb.NewError(roachpb.NewAmbiguousResultError("removing replica")),
-		ProposalRetry: proposalRangeNoLongerExists,
-	}
 	for _, p := range r.mu.proposals {
 		r.cleanupFailedProposalLocked(p)
-		p.finishApplication(pr)
+		// NB: each proposal needs its own version of the error (i.e. don't try to
+		// share the error across proposals).
+		p.finishApplication(proposalResult{
+			Err:           roachpb.NewError(roachpb.NewAmbiguousResultError("removing replica")),
+			ProposalRetry: proposalRangeNoLongerExists,
+		})
 	}
 }
 

